### ISSUE TITLE
Compilation fails on macOS if -DOPENMP_ENABLE=ON is used

### ISSUE THREAD

**Issue by raymondlanger** [LLNL/sundials#47]
The compilation fails on macOS if OpenMP is enabled:

~~~~~~~~~~~{sh}
$ cmake --version
cmake version 3.18.0
CMake suite maintained and supported by Kitware (kitware.com/cmake).
$ git clone git@github.com:LLNL/sundials.git
$ cd sundials && mkdir build && cd build
$ cmake ../ -DOPENMP_ENABLE=ON
-- The C compiler identification is AppleClang 11.0.0.11000033
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of int64_t
-- Check size of int64_t - done
-- Using int64_t for indices
-- Looking for _POSIX_TIMERS
-- Looking for _POSIX_TIMERS - not found
-- Found OpenMP_C: -Xclang -fopenmp (found version "3.1")
-- Found OpenMP: TRUE (found version "3.1")
-- Added NVECTOR_SERIAL module
-- Added NVECTOR_MANYVECTOR module
-- Added NVECTOR_OPENMP module
-- Added SUNMATRIX_BAND module
-- Added SUNMATRIX_DENSE module
-- Added SUNMATRIX_SPARSE module
-- Added SUNLINSOL_BAND module
-- Added SUNLINSOL_DENSE module
-- Added SUNLINSOL_PCG module
-- Added SUNLINSOL_SPBCGS module
-- Added SUNLINSOL_SPFGMR module
-- Added SUNLINSOL_SPGMR module
-- Added SUNLINSOL_SPTFQMR module
-- Added SUNNONLINSOL_NEWTON module
-- Added SUNNONLINSOL_FIXEDPOINT module
-- Added ARKODE module
-- Added CVODE module
-- Added CVODES module
-- Added IDA module
-- Added IDAS module
-- Added KINSOL module
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/rlanger/Downloads/tmp/sundials_test/sundials/build
Execution time: 0h:00m:05s sec
rlanger on Raymonds-MBP-2.fritz.box in ~/Downloads/tmp/sundials_test/sundials/build(75d21h36m|master*)
$ make -j16
Scanning dependencies of target sundials_generic_static_obj
Scanning dependencies of target sundials_generic_shared_obj
Scanning dependencies of target sundials_nvecserial_shared
Scanning dependencies of target sundials_nvecmanyvector_shared
Scanning dependencies of target sundials_nvecserial_static
Scanning dependencies of target sundials_nvecmanyvector_static
Scanning dependencies of target sundials_nvecopenmp_shared
Scanning dependencies of target sundials_nvecopenmp_static
Scanning dependencies of target sundials_sunmatrixband_shared
Scanning dependencies of target sundials_sunmatrixdense_static
Scanning dependencies of target sundials_sunmatrixband_static
Scanning dependencies of target sundials_sunmatrixdense_shared
Scanning dependencies of target sundials_sunmatrixsparse_shared
Scanning dependencies of target sundials_sunmatrixsparse_static
Scanning dependencies of target sundials_sunlinsolpcg_static
Scanning dependencies of target sundials_sunlinsolpcg_shared
[  0%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_shared.dir/nvector_serial.c.o
[  0%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_shared.dir/sunmatrix_dense.c.o
[  0%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_shared.dir/sunmatrix_sparse.c.o
[  0%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_static.dir/sunmatrix_dense.c.o
[  1%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_shared.dir/sunmatrix_band.c.o
[  2%] Building C object src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_static.dir/nvector_openmp.c.o
[  2%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_shared.dir/sunlinsol_pcg.c.o
[  2%] Building C object src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_shared.dir/nvector_openmp.c.o
[  2%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_static.dir/sunmatrix_band.c.o
[  2%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_shared.dir/nvector_manyvector.c.o
[  2%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_static.dir/sunmatrix_sparse.c.o
[  2%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_static.dir/sunlinsol_pcg.c.o
[  2%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_static.dir/nvector_manyvector.c.o
[  2%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_static.dir/nvector_serial.c.o
[  2%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_band.c.o
[  2%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_band.c.o
/Users/rlanger/Downloads/tmp/sundials_test/sundials/src/nvector/openmp/nvector_openmp.c:23:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
/Users/rlanger/Downloads/tmp/sundials_test/sundials/src/nvector/openmp/nvector_openmp.c:23:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
1 error generated.
1 error generated.
make[2]: *** [src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_shared.dir/nvector_openmp.c.o] Error 1
make[1]: *** [src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_shared.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[2]: *** [src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_static.dir/nvector_openmp.c.o] Error 1
make[1]: *** [src/nvector/openmp/CMakeFiles/sundials_nvecopenmp_static.dir/all] Error 2
[  3%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_dense.c.o
[  3%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_static.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_dense.c.o
[  4%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_static.dir/__/__/sundials/sundials_math.c.o
[  4%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_shared.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_shared.dir/__/__/sundials/sundials_math.c.o
[  4%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_static.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_shared.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_shared.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_shared.dir/__/__/sundials/sundials_nvector.c.o
[  4%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_static.dir/__/__/sundials/sundials_nvector.c.o
[  5%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_static.dir/__/__/sundials/sundials_nvector.c.o
[  6%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_static.dir/__/__/sundials/sundials_nvector.c.o
[  6%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_shared.dir/__/__/sundials/sundials_nvector.c.o
[  6%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_direct.c.o
[  7%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_static.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_static.dir/__/__/sundials/sundials_nvector.c.o
[  7%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_shared.dir/__/__/sundials/sundials_nvector.c.o
[  7%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_direct.c.o
[  7%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_shared.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_static.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_shared.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_static.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_shared.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_shared.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_static.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/nvector/manyvector/CMakeFiles/sundials_nvecmanyvector_static.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_shared.dir/__/__/sundials/sundials_matrix.c.o
[  7%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_static.dir/__/__/sundials/sundials_linearsolver.c.o
[  7%] Building C object src/nvector/serial/CMakeFiles/sundials_nvecserial_static.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_static.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_futils.c.o
[  7%] Building C object src/sunmatrix/dense/CMakeFiles/sundials_sunmatrixdense_shared.dir/__/__/sundials/sundials_math.c.o
[  7%] Linking C shared library libsundials_nvecserial.dylib
[  7%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_static.dir/__/__/sundials/sundials_math.c.o
[  7%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_futils.c.o
[  8%] Building C object src/sunmatrix/band/CMakeFiles/sundials_sunmatrixband_shared.dir/__/__/sundials/sundials_math.c.o
[  8%] Building C object src/sunmatrix/sparse/CMakeFiles/sundials_sunmatrixsparse_shared.dir/__/__/sundials/sundials_math.c.o
[  8%] Linking C shared library libsundials_nvecmanyvector.dylib
[  8%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_iterative.c.o
[  8%] Linking C static library libsundials_nvecmanyvector.a
[  8%] Linking C static library libsundials_nvecserial.a
[  8%] Linking C static library libsundials_sunmatrixdense.a
[  8%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_iterative.c.o
[  8%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_shared.dir/__/__/sundials/sundials_linearsolver.c.o
[  8%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_static.dir/__/__/sundials/sundials_iterative.c.o
[  8%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_linearsolver.c.o
[  8%] Linking C static library libsundials_sunmatrixsparse.a
[  8%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_math.c.o
[  8%] Linking C shared library libsundials_sunmatrixdense.dylib
[  8%] Linking C shared library libsundials_sunmatrixsparse.dylib
[  9%] Linking C static library libsundials_sunmatrixband.a
[  9%] Linking C shared library libsundials_sunmatrixband.dylib
[  9%] Building C object src/sunlinsol/pcg/CMakeFiles/sundials_sunlinsolpcg_shared.dir/__/__/sundials/sundials_iterative.c.o
[  9%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_linearsolver.c.o
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_math.c.o
[ 10%] Linking C static library libsundials_sunlinsolpcg.a
[ 10%] Built target sundials_nvecmanyvector_static
[ 10%] Built target sundials_nvecserial_static
[ 10%] Built target sundials_sunmatrixdense_static
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_matrix.c.o
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_matrix.c.o
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_nonlinearsolver.c.o
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_nonlinearsolver.c.o
[ 10%] Built target sundials_nvecserial_shared
[ 10%] Built target sundials_nvecmanyvector_shared
[ 10%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_nvector.c.o
[ 10%] Built target sundials_sunmatrixsparse_static
[ 11%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_nvector_senswrapper.c.o
[ 11%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_nvector.c.o
[ 11%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_nvector_senswrapper.c.o
[ 11%] Building C object src/sundials/CMakeFiles/sundials_generic_static_obj.dir/sundials_version.c.o
[ 11%] Building C object src/sundials/CMakeFiles/sundials_generic_shared_obj.dir/sundials_version.c.o
[ 11%] Linking C shared library libsundials_sunlinsolpcg.dylib
[ 11%] Built target sundials_sunmatrixband_static
[ 11%] Built target sundials_sunlinsolpcg_static
[ 11%] Built target sundials_sunmatrixsparse_shared
[ 11%] Built target sundials_sunmatrixband_shared
[ 11%] Built target sundials_sunmatrixdense_shared
[ 11%] Built target sundials_generic_static_obj
[ 11%] Built target sundials_generic_shared_obj
[ 11%] Built target sundials_sunlinsolpcg_shared
make: *** [all] Error 2
Execution time: 0h:00m:01s sec
~~~~~~~~~~~

In CMake 3.12+, I have added 

~~~~~~~~~~~{sh}
  if(TARGET OpenMP::OpenMP_C)
    target_link_libraries(sundials_nvecopenmp_shared PUBLIC OpenMP::OpenMP_C)
  endif()
~~~~~~~~~~~

to `src/nvector/openmp/CMakeLists.txt` to fix the issue. The same statement works for the static library.

I think that [linking the OpenMP target](https://cliutils.gitlab.io/modern-cmake/chapters/packages/OpenMP.html) is the idiomatic way of enabling OpenMP in CMake, but I have not tested this for older CMake versions or `OPENMP_DEVICE_ENABLE=ON`.

**Comment by balos1** [comment]
Thanks for reporting this issue. The next version of SUNDIALS will use the OpenMP target, so it should resolve this issue.

**Comment by balos1** [comment]
Not fixed in v5.4.0, will be fixed in the following release.

**Comment by balos1** [comment]
Should be fixed in v5.5.0, please reopen the issue if it persists.
